Show the total when paginating Field::HasMany

maxkwallace opened the pull request:
The new pagination feature added in https://github.com/thoughtbot/administrate/commit/41a5e258c109fe44e80085bc2e4d27c476fa9879 is awesome-- but as soon as the total disappeared from the UI, our users asked for it back :)

maxkwallace commented:
This is unrelated to this PR, but when attempting to run `bin/setup` and the tests with Ruby 2.4.0 I ran into the following. (Tests worked fine locally with Ruby 2.3.1.)

```
mkw@M4600:~/repos/administrate-fork/administrate-fork$ ./bin/setup
The Gemfile's dependencies are satisfied
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/sprockets-3.7.0/lib/sprockets/digest_utils.rb:47: warning: constant ::Fixnum is deprecated
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/sprockets-3.7.0/lib/sprockets/digest_utils.rb:51: warning: constant ::Bignum is deprecated
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/sprockets-3.7.0/lib/sprockets/processor_utils.rb:110: warning: constant ::Fixnum is deprecated
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/sprockets-3.7.0/lib/sprockets/processor_utils.rb:111: warning: constant ::Bignum is deprecated
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/concurrent-ruby-1.0.2/lib/concurrent/map.rb:206: warning: constant ::Fixnum is deprecated
administrate-prototype_development already exists
administrate-prototype_test already exists
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/activesupport-4.2.7.1/lib/active_support/core_ext/numeric/conversions.rb:121: warning: constant ::Fixnum is deprecated
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/activesupport-4.2.7.1/lib/active_support/core_ext/numeric/conversions.rb:121: warning: constant ::Bignum is deprecated
rake aborted!
SystemStackError: stack level too deep
/home/mkw/.rvm/gems/ruby-2.4.0@shearwater/gems/activesupport-4.2.7.1/lib/active_support/core_ext/numeric/conversions.rb:131:in `block (2 levels) in <class:Numeric>'
```

and

```
mkw@M4600:~/repos/administrate-fork/administrate-fork$ bundle exec appraisal rake
>> BUNDLE_GEMFILE=/home/mkw/repos/administrate-fork/administrate-fork/gemfiles/sass_3_4.gemfile bundle exec rake
Could not find gem 'rails-controller-testing' in any of the gem sources listed in your Gemfile.
Run `bundle install` to install missing gems.
```

tysongach commented on app/assets/stylesheets/administrate/components/_pagination.scss:
```diff
@@ -15,3 +15,15 @@
 }
+
+.pagination-with-total {
+  margin-bottom: $base-spacing * 4;
```
I don’t think we need this `margin-bottom` here, do we? `.pagination` already has the same `margin-bottom` applied on Line 2.

maxkwallace replied:
@tysongach the HTML looks like:

```
<div class="pagination--with-total">
  <nav class='pagination'>
    ...
  </nav>
  <div class="pagination__total">
    ...
  </div>
</div>
```

The pagination nav is, AFAIK, generated by Kaminari, and there's no way to customize that bit of markup without overriding Kaminari's default pagination helper, which seems out of scope for this PR.

I think we do need this `margin-bottom`-- because `.pagination--with-total` is a wrapper around `.pagination`, the outer div should have the same margin bottom in the end.

tysongach commented on app/assets/stylesheets/administrate/components/_pagination.scss:
```diff
@@ -15,3 +15,15 @@
   }
 }
+
+.pagination-with-total {
```
We utilize [BEM class naming](https://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/). Given that `.pagination` is the block, other styles within this partial should be elements and modifiers. For example, it might make sense for this to be a modified version of the `pagination` class, which would dictate this class as `pagination--with-total`.

We also try and avoid unnecessary nesting and instead maintain a flat, low-specificity CSS architecture.

maxkwallace replied:
Thanks for the comment!

I updated the class to `pagination--with-total` and de-nested the `pagination__total` style. LMK if that doesn't look better!

tysongach commented on app/assets/stylesheets/administrate/components/_pagination.scss:
```diff
@@ -15,3 +15,15 @@
+.pagination-with-total {
+  margin-bottom: $base-spacing * 4;
+
+  .pagination {
```
Is this needed? Can we simply reuse the `margin-bottom` already applied to `.pagination` on Line 1?

maxkwallace replied:
@tysongach Without this style, there wouldn't the right amount of space between the pagination links and text indicating the total.

tysongach commented on app/assets/stylesheets/administrate/components/_pagination.scss:
```diff
@@ -15,3 +15,15 @@
+  }
+
+  .pagination-total {
+    text-align: center;
```
Is this needed? The `.pagination` on Line 1 already sets this, and it is the parent to `. pagination-total`. Seems like we should be able to remove this class all-together.

maxkwallace replied:
@tysongach `.pagination` isn't the parent of `pagination__total`. This style is needed.

maxkwallace commented:
![before](https://cloud.githubusercontent.com/assets/3249377/24957700/86a04e5a-1f5a-11e7-8c10-095811e10ae7.png)
![after](https://cloud.githubusercontent.com/assets/3249377/24957699/86761284-1f5a-11e7-8b86-349d7a0c7e76.png)

maxkwallace commented:
@tysongach I added some docs to CONTRIBUTING.md in an attempt to help alleviate a couple issues I ran into when trying to test this again.

maxkwallace commented:
@tysongach AFAIK I've addressed all your comments-- please re-review when you get a chance. I've also rebased this fork on master and re-squashed my commits.

I don't think the test failures have to do with anything I've added in this PR. `bundle exec rake` seems to be working fine, but `bundle exec appraisal rake` seems to be failing 

locally:

> Bundler could not find compatible versions for gem "activesupport":
  In Gemfile:
    rails-controller-testing was resolved to 1.0.1, which depends on
      actionpack (~> 5.x) was resolved to 5.0.2, which depends on
        activesupport (= 5.0.2)

and in CircleCI:

> bundle exec appraisal rake
Ignoring json-1.8.6 because its extensions are not built.  Try: gem pristine json --version 1.8.6
Ignoring nio4r-1.2.1 because its extensions are not built.  Try: gem pristine nio4r --version 1.2.1
Ignoring nokogiri-1.7.0.1 because its extensions are not built.  Try: gem pristine nokogiri --version 1.7.0.1
Ignoring nokogiri-1.6.8.1 because its extensions are not built.  Try: gem pristine nokogiri --version 1.6.8.1
Ignoring nokogiri-1.6.8 because its extensions are not built.  Try: gem pristine nokogiri --version 1.6.8
Ignoring pg-0.19.0 because its extensions are not built.  Try: gem pristine pg --version 0.19.0
Ignoring raindrops-0.17.0 because its extensions are not built.  Try: gem pristine raindrops --version 0.17.0
Ignoring unicorn-5.2.0 because its extensions are not built.  Try: gem pristine unicorn --version 5.2.0
Ignoring websocket-driver-0.6.3 because its extensions are not built.  Try: gem pristine websocket-driver --version 0.6.3
>> BUNDLE_GEMFILE=/home/ubuntu/administrate/gemfiles/sass_3_4.gemfile bundle exec rake
Ignoring json-1.8.6 because its extensions are not built.  Try: gem pristine json --version 1.8.6
Ignoring nio4r-1.2.1 because its extensions are not built.  Try: gem pristine nio4r --version 1.2.1
Ignoring nokogiri-1.7.0.1 because its extensions are not built.  Try: gem pristine nokogiri --version 1.7.0.1
Ignoring nokogiri-1.6.8.1 because its extensions are not built.  Try: gem pristine nokogiri --version 1.6.8.1
Ignoring nokogiri-1.6.8 because its extensions are not built.  Try: gem pristine nokogiri --version 1.6.8
Ignoring pg-0.19.0 because its extensions are not built.  Try: gem pristine pg --version 0.19.0
Ignoring raindrops-0.17.0 because its extensions are not built.  Try: gem pristine raindrops --version 0.17.0
Ignoring unicorn-5.2.0 because its extensions are not built.  Try: gem pristine unicorn --version 5.2.0
Ignoring websocket-driver-0.6.3 because its extensions are not built.  Try: gem pristine websocket-driver --version 0.6.3
bundler: failed to load command: rake (/home/ubuntu/administrate/vendor/bundle/ruby/2.3.0/bin/rake)
Gem::LoadError: You have already activated bundler 1.13.7, but your Gemfile requires bundler 1.14.6. Prepending `bundle exec` to your command may solve this.
  /opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.6/lib/bundler/runtime.rb:40:in `block in setup'
  /opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.6/lib/bundler/runtime.rb:25:in `map'
  /opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.6/lib/bundler/runtime.rb:25:in `setup'
  /opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.6/lib/bundler.rb:100:in `setup'
  /opt/circleci/.rvm/gems/ruby-2.3.1/gems/bundler-1.14.6/lib/bundler/setup.rb:10:in `<top (required)>'
  /opt/circleci/ruby/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
  /opt/circleci/ruby/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'

bundle exec appraisal rake returned exit code 1

nickcharlton commented:
Hi @maxkwallace! Could I get you to rebase again? I wonder with the time that's past since your issues might have gone away.

maxkwallace commented:
@nickcharlton just rebased (no merge conflicts, so the code in this PR hasn't changed). `rake` passes for me locally.

If you don't think this is mergeable, could you clarify why? 

(The test failure in CI isn't related to the code in this branch, and I've tried running `rake TESTOPTS="--seed=931"` locally, but strangely, rake isn't picking up the --seed flag.)

nickcharlton commented:
Thanks! The failure you're seeing is related to #846.

I'll have a proper look shortly, but it looks like we're good to go.

maxkwallace commented:
@nickcharlton hey Nick-- just wanted to check in, due to the 'waiting-on-response' label. Is anyone still waiting on me for anything?

I just rebased again.

maxkwallace commented:
@nickcharlton it's been ~6 weeks. Any update on this?

nickcharlton commented:
Oh, thanks! This had slipped me by.

Looking at this properly this time, I wonder if we could put it all on one line? Perhaps:

> Prev 1 2 3 … 10 Next Last

What do you think? Perhaps that approach isn't quite clear enough.

maxkwallace commented:
@nickcharlton I'm fine with that change, but I'd rather get this merged in first, if possible, and then iterate on the feature once it's back in, if that ends up being a priority for the project. This PR has been open since March 30th-- having WIP sitting around that long isn't ideal, so let's merge it or just close it.

maxkwallace commented:
@tysongach thanks for the comment! Re: what I posted before:

> let's merge it or just close it.

I'm closing this. If you folks want to fork my fork, open a new PR, and iterate on it there, you're more than welcome.
